src/ir/possible-contents.cpp

```diff
@@ -1321,12 +1323,504 @@ struct InfoCollector
+                                  TNHInfo& info) {
+  // Optimize in the same basic block as the call: all instructions still in
+  // that block will definitely execute if the call is reached. We will do that
+  // by going backwards through the call's operands and fallthrough values, and
+  // optimizing while we are still in the same basic block.
+  auto callBlockIndex = blockIndexes.get(call);
+
+  // Operands must exist since there is a cast param, so a param exists.
+  assert(operands.size() > 0);
+  for (int i = int(operands.size() - 1); i >= 0; i--) {
+    auto* operand = operands[i];
+
+    if (blockIndexes.get(operand) != callBlockIndex) {
+      // Control flow might transfer; stop.
+      break;
+    }
+
+    auto iter = targetCastParams.find(i);
+    if (iter == targetCastParams.end()) {
+      // This param is not cast, so skip it.
+      continue;
+    }
+
+    // If the call executes then this parameter is definitely reached (since it
+    // is in the same basic block), and we know that it will be cast to a more
+    // refined type.
+    auto castType = iter->second;
+
+    // Apply what we found to the operand and also to its fallthrough
+    // values.
+    //
+    // At the loop entry |curr| has been checked for a possible control flow
+    // transfer (and that problem ruled out).
+    auto* curr = operand;
+    bool transferred = false;
+    while (1) {
+      // Note the type if it is useful.
+      if (castType != curr->type) {
+        // There are two constraints on this location: any value there must
+        // be of the declared type (curr->type) and also the cast type, so
+        // we know only their intersection can appear here.
+        auto declared = PossibleContents::fullConeType(curr->type);
+        auto intersection = PossibleContents::fullConeType(castType);
+        intersection.intersect(declared);
+        if (intersection.isConeType()) {
+          auto intersectionType = intersection.getType();
+          if (intersectionType != curr->type) {
+            // We inferred a more refined type.
+            info.inferences[curr] =
+              PossibleContents::fullConeType(intersectionType);
+          }
+        } else {
+          // The intersection of two full cones must be another cone, which we
+          // handled above, or must be empty, which we handle here.
+          assert(intersection.isNone());
+
+          // Nothing is possible here, so this must be unreachable code.
+          info.inferences[curr] = PossibleContents::none();
+        }
+      }
+
+      auto* next = Properties::getImmediateFallthrough(curr, options, wasm);
+      if (next == curr) {
+        // No fallthrough, we're done with this param.
+        break;
+      }
+
+      // There is a fallthrough. Check for a control flow transfer.
+      if (blockIndexes.get(next) != callBlockIndex) {
+        // Control flow might transfer; stop.
+        transferred = true;
+        break;
+      }
```

Comment:
Can we move this out to where we check `transferred` anyway below?

Reply:
In general we'd need a labelled break here (does C++ have those?) since we are nested inside a loop here and must break out of it first in order to reach the lower break. But I realize that we can just return from the entire function anyhow... I'll do that.